New: `prefer-destructuring` rule (fixes #6053)

alexlafroscia opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[x] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**Please describe what the rule should do:**

This rule ensures that all object property access and array index accesses are done through [ES6 destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment).

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[ ] Warns about a potential error
[x] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```diff
- var foo = array[0];
+ var [ foo ] = array;

- var foo = object.foo;
+ var { foo } = object;

- var foo = object['foo'];
+ var { foo } = object;
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This change supports (an encourages) a new JavaScript feature that isn't tied directly to any particular framework or library. Additionally, it _is_ part of an existing plugin, [eslint-plugin-ember-suave](https://github.com/DockYard/eslint-plugin-ember-suave/blob/master/lib/rules/prefer-destructuring.js), where I originally authored the rule.  However, someone requested that the rule be added to core in #6053 and asked if I would submit it to core.

**What changes did you make? (Give an overview)**

- Added a new rule, `prefer-destructuring`, as well as the tests and documentation that I originally authored as part of the `eslint-plugin-ember-suave`.

**Is there anything you'd like reviewers to focus on?**

Please make sure the tests and documentation are thorough enough. I just copied the implementation, tests and docs from the existing plugin (linked above), which may not have the same level of standards that the core rules have, although I did my best to follow all the best practices when I originally wrote the rule.

not-an-aardvark commented on lib/rules/prefer-destructuring.js:
```diff
@@ -0,0 +1,128 @@
+                const variableIdentifier = node.id;
+                const property = node.init.property;
+
+                if (property.type === "Literal" && variableIdentifier.name === property.value) {
```
It looks like this only reports

```js
var foo = bar.foo;
```

but it doesn't report

```js
var foo = bar.baz;
```

I think this is incorrect behavior, because the latter could be replaced with

```js
var {baz: foo} = bar;
```

alexlafroscia replied:
Good point. I think it would be good to solicit some feedback on this. My assumption was that, if the property name and variable name didn't match, the rule just isn't applicable. However, we're right that it _could_ be done the other way, although writing out

```js
var { baz: foo } = bar;
```

is way more confusing to read than

```js
var foo = bar.baz;
```

I'm not sure the rule should enforce the first destructure-and-alias-name version; maybe that should be a configuration option?

not-an-aardvark replied:
I'd also be fine with having a configuration option for it.

alexlafroscia replied:
So, that would make the configuration options look something like:

```json
{
  "rules": {
    "prefer-destructuring": ["error", {
      "array": true,
      "object": true
    }]
  }
}
```

to exhibit the current behavior (not requiring that mis-matched names be destructured) and then something like

```json
{
  "rules": {
    "prefer-destructuring": ["error", {
      "array": true,
      "object": { "some-configuration-key": true }
    }]
  }
}
```

for the more strict approach?  What might be a good name for `some-configuration-key`?

alexlafroscia replied:
Or, would the more-strict definition be the default, with an option to make it more relaxed?

alexlafroscia replied:
The refactoring I did here makes it seem like this was address (because the code changed), but it hasn't been yet.

not-an-aardvark replied:
I had been thinking it would be a separate option,  along the lines of

```json
{
  "rules": {
    "prefer-destructuring": ["error", {
      "array": true,
      "object": true
    }, {
      "requireRenaming": false
    }]
  }
}
```

But I'm definitely open to bikeshedding on the option name/configuration.

---

In general, I slightly prefer that rules have strict behavior by default, and allow the user to manually relax the behavior. My reasoning is that a user might want to just turn on the rule before carefully considering the available options. If the rule is strict by default and the user preferred the loose setting, the issue will probably be obvious (errors will get reported, and the user will look up the options and relax the behavior). On the other hand, if the rule is loose by default and the user preferred the strict setting, the user might not notice anything is wrong, and they'll think ESLint is validating something when it's actually not.

alexlafroscia replied:
RE: strict by default, I think that sounds like a great justification for going that route.

I'll take a look at implementing that more-strict behavior as optional ASAP, based on the configuration that you demonstrated. If someone proposes a syntax that we like better, that should be easy enough to change.

alexlafroscia replied:
Add the configuration suggested above and some tests to cover the it.

not-an-aardvark commented on lib/rules/prefer-destructuring.js:
```diff
@@ -0,0 +1,128 @@
+        //--------------------------------------------------------------------------
+
+        return {
+            VariableDeclarator: checkVariableDeclarator
```
It looks like this only checks `VariableDeclarator` nodes, but I think it should also check `AssignmentExpression` nodes.

For example, the following should get reported as an error:

```js
foo = bar[0];
```

Since it could get replaced with:

```js
[foo] = bar;
```

This should also get reported:

```js
foo = bar.baz;
```

because it could be replaced with:

```js
({baz: foo} = bar);
```

not-an-aardvark commented on tests/lib/rules/prefer-destructuring.js:
```diff
@@ -0,0 +1,100 @@
+// Tests
+//------------------------------------------------------------------------------
+
+const ruleTester = new RuleTester();
```
Nit: Instead of repeating `{ parserOptions: 6 }` for each test, it would be better to pass this as a default to the `RuleTester` constructor.

```js
const ruleTester = new RuleTester({ parserOptions: { ecmaVersion: 6 } });
```

not-an-aardvark commented on lib/rules/prefer-destructuring.js:
```diff
@@ -0,0 +1,128 @@
+         * @returns {boolean} whether or not the node is an integer
+         */
+        function isArrayIndexAccess(node) {
+            return Number.isInteger(node.property.value);
```
This check looks fine and it's correct behavior, but as far as usability is concerned, I'm thinking it might be annoying to get a warning for code like this:

```js
const foo = array[100];
```

The correct fix for that code would be:

```js
const [ , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , foo] = array;
```

alexlafroscia replied:
I added a note to the docs in the "When not to use this" section about accessing large indices directly, and that not playing very nicely with destructuring. Now sure what else we could do, other than setting some arbitrary (or user-configurable) threshold on when to require it.

alexlafroscia commented on lib/rules/prefer-destructuring.js:
```diff
@@ -0,0 +1,128 @@
+                    const value = prop.value;
+
+                    if ((key.name === value.name) && (key.start !== value.start)) {
+                        context.report({ node: prop, message: "Unnecessary duplicate variable name" });
```
Yes, you're right that that's what that intends to protect against. Assuming `object-shorthand` would catch that, I can remove the check here.

not-an-aardvark commented on lib/rules/prefer-destructuring.js:
```diff
@@ -0,0 +1,165 @@
+            }
+
+            if (checkArrays && isArrayIndexAccess(rightNode)) {
+                context.report({ node: reportNode, message: "Use array destructuring" });
```
Minor nit: Can you extract the `context.report` calls into a separate helper function (e.g. `reportNode`)? In my opinion, repeating the same report message four times in the code is a bit undesirable.

not-an-aardvark commented on lib/rules/prefer-destructuring.js:
```diff
@@ -0,0 +1,165 @@
+            if (checkObjects && !isArrayIndexAccess(rightNode)) {
+                const property = rightNode.property;
+
+                if (requireRenaming && !rightNode.computed) {
```
I don't think the `!rightNode.computed` check is necessary here; computed properties aren't any different from other properties in this case:

```js
const foo = bar[baz];

// can be replaced with

const {[baz]: foo} = bar;
```

alexlafroscia replied:
Oh wow, I had no idea that that syntax was valid. That is.... pretty weird. But, you're right!

not-an-aardvark commented on docs/rules/prefer-destructuring.md:
```diff
@@ -0,0 +1,61 @@
+
+### Options
+
+This rule takes two properties, `array` and `object`, which can be used to turn on or off the destructuring requirement for each of those types independently.  By default, both are `true`.  If you want to change the behavior, you can configure the rule like so:
```
We should also add documentation for the `requireRenaming` option (although it would also be fine to wait and see if we plan to bikeshed on the name).

not-an-aardvark commented:
I'll repost the discussion here for greater visibility, since the [original discussion](https://github.com/eslint/eslint/pull/7741#discussion_r91868522) is hidden under an "outdated diff" thing.

Basically, there is an option to toggle whether or not the rule should require destructuring in cases where properties would need to be renamed.

```js
const foo = bar.baz;

// this could be replaced with 
const {baz: foo} = bar;
```

We should decide:

* What should the option be called? (Currently, it's called `requireRenaming`)
* What should the rule do by default? (Currently, the rule *does* require renaming by default. I argued [here](https://github.com/eslint/eslint/pull/7741#discussion_r92036721) for the strict-by-default behavior, but I'm starting to reconsider this because the second line above seems significantly less readable than the first line.)

kaicataldo commented:
> What should the option be called? 
I think we should avoid `require` in this option name since it's kind of a loaded term (not a big deal, but I think it could be clearer to use something like `prefer` or `enforce`).

> What should the rule do by default?
My preference would be to leave the option off (should not enforce destructuring with renaming), because I don't think that's a common pattern and agree that it's harder to read than the first.

ljharb commented:
How about `enforceForRenamedProperties`? (definitely defaulted to off)

alexlafroscia commented:
Alright, I updated the docs, renamed the additional configuration option to `enforceForRenamedProperties`, made it off by default and required that crazy 

```javascript
var { [bar]: foo } = object;
```

syntax if it is enabled.

ljharb commented on docs/rules/prefer-destructuring.md:
```diff
@@ -23,18 +23,39 @@ var [ foo ] = array;
 
 // With `object` enabled
 var { foo } = object;
+var foo = object.bar
```
`;`?

ljharb commented on tests/lib/rules/prefer-destructuring.js:
```diff
@@ -83,15 +93,23 @@ ruleTester.run("prefer-destructuring", rule, {
         {
-            code: "var foo = array.bar;",
-            options: [{ object: true }, { requireRenaming: true }],
+            code: "var foo = object[bar];",
```
i'm not sure i understand - `object[bar]` isn't the same as `object.bar`. How would you use destructuring here? Does `var { [bar]: foo } = object;` work?

not-an-aardvark replied:
Apparently it does, yes.

ljharb replied:
how do i javascript, 2016 edition

not-an-aardvark replied:
For bonus Readable Code™ points, figure out what this function does!

```js
(({[foo = 1]:[foo = 2]} = {[foo = 3]:[foo = 4]}) => ({[foo = 5]:[foo = 6]}))
```

ljharb replied:
`const z̲̗̼͙̥͚͛͑̏a̦̟̳͋̄̅ͬ̌͒͟ļ̟̉͌ͪ͌̃̚g͔͇̯̜ͬ̒́o̢̹ͧͥͪͬ =`

kaicataldo replied:
Huh...TIL

ilyavolodin commented on docs/rules/prefer-destructuring.md:
```diff
@@ -0,0 +1,82 @@
+# Prefer destructuring from arrays and objects (prefer-destructuring)
+
+With JavaScript ES6, a new syntax was added for creating variables from an array index of object property, called [destructuring](#further-reading).  This rule enforces usage of destructuring instead of accessing a property through a member expression.
```
Should this be "from an array index **or** object property"?

alexlafroscia replied:
Yup, thanks!

ilyavolodin commented on docs/rules/prefer-destructuring.md:
```diff
@@ -0,0 +1,82 @@
+
+With JavaScript ES6, a new syntax was added for creating variables from an array index of object property, called [destructuring](#further-reading).  This rule enforces usage of destructuring instead of accessing a property through a member expression.
+
+## Rule Details
```
Can you add a paragraph that array destructuring is only enforced if array index is a literal? The last "correct" example is slightly confusing without this note. As in, this rule will flag this code: `var a = array[100];` but will not flag this code: `var a = array[myVar];`

alexlafroscia replied:
What about just expanding the list of "correct" examples to look like:

```javascript
// With `array` enabled
var [ foo ] = array;
var foo = array[someIndex];
```

ilyavolodin replied:
That sounds goo too.

not-an-aardvark commented:
Thanks for contributing to ESLint!

kaicataldo commented:
@alexlafroscia Appreciate you being so willing to add this when we asked!
